[RFC] add mux/demux documentation

juimonen opened the pull request:
Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

lgirdwood commented on algos/demux/demux.rst~:
```diff
@@ -0,0 +1,78 @@
+.. _demux:
+
+Muliplexer/Demultiplexer
+########################
```
This looks like a duplicate of the previous file ?

juimonen replied:
oh, emacs backup, sorry will remove.

juimonen commented:
@lgirdwood updated. 
- removed emacs backup file
- removed white space
- fixed a typo in header

This is now depending on: 
- https://github.com/thesofproject/sof/pull/2594 
- https://github.com/thesofproject/sof/pull/2600.

I also talk about python blob generator: https://github.com/thesofproject/sof/pull/2536, which people seem not to like, but it is the only "generic" tool we would have to generate blobs... But I can remove this if no traction. So the python tool could be used to generate dc blocker, mux/demux, or even simple eq blobs.

And now when I'm on it: we should really do the optimization for demux @Keyon was talking about in https://github.com/thesofproject/sof/issues/2533.

Then the demux would be in pretty good shape, so this here looks like mux todo list :)

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+
+.. figure:: images/mux.png
+
+   Example of multiplexer configuration matrises with 2 input streams.
```
matrises -> matrices?

ranj063 replied:
or maybe just matrix?

juimonen replied:
yes

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+
+.. figure:: images/demux.png
+
+   Example of demultiplexer configuration matrises with 2 output streams.
```
ditto

juimonen replied:
yes

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+========
+
+Previous figures show that the routing matrix is difficult to
+parametrize to be easily understandable. As it is send to firmware
```
send->sent?

juimonen replied:
yes

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+to "visualize" the matrix for easier configuration.
+
+Example from pipe-volume-demux-playback.m4 shows how to define 2
+routing matrises and a demux component:
```
matrices

juimonen replied:
yes

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+Alsa control
+============
+
+Multiplexer configuration is loaded in kernel/firmware boot as part of Alsa
```
ALSA

juimonen replied:
ok

ranj063 commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+============
+
+Multiplexer configuration is loaded in kernel/firmware boot as part of Alsa
+binary control in topology, but can be also controlled through alsa
```
ALSA all caps

juimonen replied:
ok

juimonen commented:
@ranj063 thanks for comments! updated now, hopefully addressed all comments...

libinyang commented on algos/demux/demux.rst:
```diff
@@ -0,0 +1,138 @@
+   Example of multiplexer configuration matrices with 2 input streams.
+   In this artificial Mux example Input stream 1’s channel 1 is copied
+   to both output channels, Input stream 2’s single channel is copied
+   also to output channel 1.
```
@juimonen I'm a newer to mux/demux. And I find it's a little hard for me to understand to understand the picture. Why do we need set column 3, 5, 7 to 1? Shouldn't it be 0s although it doesn't take efforts here. Or we must set 1s in column 3, 5, 7?

juimonen replied:
@libinyang just think as the row as input channel, and 0/1 as if this row (channel) is put to column output channel. so for example first row (0,1,0,1,0,1,0,1) would mean that copy input channel 1 to output channels 2, 4, 6, 8. And so on for other channels. There's always 1 matrix per stream-to-stream mapping.

keyonjie commented:
> @lgirdwood updated.
> 
> * removed emacs backup file
> * removed white space
> * fixed a typo in header
> 
> This is now depending on:
> 
> * [thesofproject/sof#2594](https://github.com/thesofproject/sof/pull/2594)
> * [thesofproject/sof#2600](https://github.com/thesofproject/sof/pull/2600).
> 
> I also talk about python blob generator: [thesofproject/sof#2536](https://github.com/thesofproject/sof/pull/2536), which people seem not to like, but it is the only "generic" tool we would have to generate blobs... But I can remove this if no traction. So the python tool could be used to generate dc blocker, mux/demux, or even simple eq blobs.
> 
> And now when I'm on it: we should really do the optimization for demux @Keyon was talking about in [thesofproject/sof#2533](https://github.com/thesofproject/sof/issues/2533).
> 
> Then the demux would be in pretty good shape, so this here looks like mux todo list :)

Hi @juimonen, thanks for the documentation! Only one suggestion that to remove the complicate matrix part if we are aligned to remove the "channel mixing" feature in the future. The good point is that nobody is using that mixing feature until now, so don't mention that could be wise choice, to make easier life not only for people reading this documentation, but also for people who might need to maintain/update the documentation later.

deb-intel commented:
@lgirdwood Thanks!
